### ISSUE TITLE
Howto Use interpolation format function with parameters?

### ISSUE THREAD

**Issue by daKmoR** [i18next/i18next#992]
hey,

so we have a ton of already translated elements in the Intl Api but I really would like to move to i18next.
Problem is some of those translation messages are using pretty specific Intl Apis. (e.g. intl-messageformat)

I though yeah no problem just provide interpolate format function... but I don't have the parameters in the interpolation format function...

If I had I could just add a format for Intl like so right?
```
en:
  translations:
    key: "Hey {{there}}",
    keyWithIntl: "{{Hey {there}, Intl}}"

interpolation:
  format: function(value, format, language, parameters) { //parameters would be new
    if (format === 'Intl') {
      let intlObject= new IntlMessageFormat(value, language);
      return intlObject.format(parameters);
    }
  }
```

**Comment by jamuhl** [comment]
i personally already thought about adding a translator supporting icu format. But ended in not doing it...as there are a lot of options out there to support icu. But did not have the migration case in mind.

But for a migration use case i would not do it by interpolation - but adding a post processor: https://www.i18next.com/misc/creating-own-plugins.html#post-processor

using that like eg. the interval plural post processor: https://github.com/i18next/i18next-intervalPlural-postProcessor

so instead of using format, which only passes an interpolate value you get the complete found value - translation in message format and from there you can parse and format like in your code snipplet.

**Comment by daKmoR** [comment]
oh I see - it's a pity it does not work as I hoped... would have been so easy :p but makes total sense :)
never mind as long as there is a way I guess I can make it work - so I'm off trying my luck on creating an intl postProcessor :p
thx for the hint and the example :)

**Comment by daKmoR** [comment]
Simple Test
```
let IntlPostProcessor = {
  type: 'postProcessor',
  name: 'intl',
  process: function(value, key, options, translator) {
    console.log(value, key, options, translator);
    return value;
  }
};

i18next
  .use(i18nextXHRBackend)
  .use(i18nextBrowserLanguageDetector)
  .use(IntlPostProcessor)
  .init(settings);
````

console.log never get's called? what am I missing?

**Comment by daKmoR** [comment]
ok the PostProcessor was actually easy :p

```
let IntlPostProcessor = {
  type: 'postProcessor',
  name: 'intl',
  process: function(value, key, options, translator) {
    if (value.indexOf('{') !== -1 && value.indexOf('{{') === -1) {
      let parameters = options[1];
      let intlObject = new IntlMessageFormat(value, translator.language);
      return intlObject.format(parameters);
    }
    return value;
  }
};
```

but howto add it by default to all the translations?
the following init option seems to only work for messages that have no options?
```
      overloadTranslationOptionHandler: function(config) {
        return Object.assign({}, config, {
          postProcess: 'intl'
        });
      },
```

also as seen here
https://github.com/i18next/i18next/blob/master/src/i18next.js#L233

it will only be used if there are no options? but I want to always add a default option? any other option available?

**Comment by daKmoR** [comment]
OMG...  if I could just read :p there is an option postProcess

so just setting `postProcess: ['intl']` and it works just fine with

```
let IntlPostProcessor = {
  type: 'postProcessor',
  name: 'intl',
  process: function(value, key, options, translator) {
    if (value.indexOf('{') !== -1 && value.indexOf('{{') === -1) {
      let intlObject = new IntlMessageFormat(value, translator.language);
      return intlObject.format(options);
    }
    return value;
  }
};
```

maybe adding a simple demo on howto do this in the docs would be nice?

should I create a PR for gitbook?

**Comment by jamuhl** [comment]
https://www.i18next.com/configuration-options.html#translation-defaults -> init options.postProcess (array of postProcessors to use per default)

or default options: https://www.i18next.com/interpolation.html#all-interpolation-options -> defaultVariables (can be added on init too)

for adding to the docs...yes...feel free to add it to the docs...PR welcome...maybe a new section under misc?

**Comment by tonix-tuft** [comment]
I have a use case where I need to use the parameters/options provided to the `t` function when formatting (I am not using `Intl`).

Is there a way to do that as of today?

E.g.:

```
// JSON File:
{
    ...
    keyWithFormatting: "{number,myFormat}}",
    ...
}

// Usage:
i18next.t("keyWithFormatting", { number: 999, param1: 123, param2: 456 });
...
interpolation:
  format: function(value, format, language, options) { // 
    if (format === 'myFormat') {
      // "value" is 999 here. But I need to access "param1" and "param2" as well because my 
      // formatting depends on this options too and they are dynamic...
      // Is there a way to do that?
      ...
    }
  }
...
```

**Comment by jamuhl** [comment]
@tonix-tuft  can be added here: https://github.com/i18next/i18next/blob/master/src/Interpolator.js#L87 do you like to provide a PR?

**Comment by tonix-tuft** [comment]
Would love to do that. Could you please instruct me on how I can do it? It would be my first contribution/pull request...

Thank you.

**Comment by jamuhl** [comment]
@tonix-tuft hope that helps: https://akrabat.com/the-beginners-guide-to-contributing-to-a-github-project/ should be rather straight forward: fork, change, push, create PR

**Comment by tonix-tuft** [comment]
Will take a look, try it and let you know thanks!

**Comment by jamuhl** [comment]
@tonix-tuft had time today - landed in i18next@19.1.0 -> 

```js
interpolation:
  format: function(value, format, language, options) { } // newly added options as 4th argument
```

**Comment by tonix-tuft** [comment]
That's awesome, I was going to take a look at it this afternoon. Thank you very much!
